core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java

```diff
@@ -41,88 +41,128 @@
  * to destination database
  */
 public class FileCopyRecordWriter extends RecordWriter<NullWritable, AbstractFileMetadata> {
-  private FileSystem destFileSystem;
-  private String basePath;
-  private Boolean enableOverwrite;
-  private int bufferSize;
+  private final FileSystem destFileSystem;
+  private final String basePath;
+  private final boolean enableOverwrite;
+  private final boolean preserveOwner;
+  private final boolean fsCache;
+  private final int bufferSize;
+
   // buffer size defaults to 1 MB
   private static final int defaultBufferSize = 2 << 20;
   private static final Logger LOG = LoggerFactory.getLogger(FileCopyRecordWriter.class);
 
-  public FileCopyRecordWriter(Configuration conf)
-    throws Exception {
-    this.destFileSystem = FileSystem.get(conf);
-    this.basePath = conf.get(FileCopyOutputFormat.BASE_PATH);
-    this.enableOverwrite = conf.getBoolean(FileCopyOutputFormat.ENABLE_OVERWRITE, false);
-    this.bufferSize = conf.getInt(FileCopyOutputFormat.BUFFER_SIZE, defaultBufferSize);
-    this.sourceFilesystemMap = new HashMap<>();
-  }
-
   // source filesystems
-  private Map<AbstractFileMetadata.Credentials, FileSystem> sourceFilesystemMap;
+  private Map<String, FileSystem> sourceFilesystemMap;
 
-  @Override
-  public void write(NullWritable key, AbstractFileMetadata fileMetadata) throws IOException, InterruptedException {
-    // get source database connection
-    AbstractFileMetadata.Credentials sourceCredentials = fileMetadata.getCredentials();
-    if (!sourceFilesystemMap.containsKey(sourceCredentials)) {
-      sourceFilesystemMap.put(sourceCredentials, getSourceFilesystemFromCredentials(sourceCredentials));
+  public FileCopyRecordWriter(Configuration conf) throws Exception {
```

Comment:
should throw as specific an exception as possible. I think this should be IOException.